File: DP(Dynamic_Programming, 동적계획법)/백준_2748_피보나치 수2/피보나치수2.java
import java.io.*;
import java.util.*;

public class Main {
    static long[] d = new long[100];

    public static long dp(int n){
        //1. base check
        if(n <= 1) return n;

        //2. memoization
        if(d[n] > 0) return d[n];

        //3. recurrence relation
        d[n] = dp(n-1) + dp(n-2);

        return d[n];
    }

    public static void main(String[] args) throws IOException {
        Scanner s = new Scanner(System.in);
        int n = s.nextInt();
        System.out.println(dp(n));

    }
}

File: DP(Dynamic_Programming, 동적계획법)/백준_2748_피보나치 수2/피보나치수2.kt
import java.util.*

var d = LongArray(100, {0})

fun dp(n: Int):Long {
    if(n == 0) return 0
    if(n == 1) return 1

    if(d[n] > 0) return d[n]

    d[n] = dp(n-1) + dp(n-2)

    return d[n]
}

fun main(args: Array<String>) {
    val s = Scanner(System.`in`)
    val input = s.nextInt()
    print("${dp(input)}")
}